### PR TITLE
[12.0] backport l10n_ch_qr_bill_scan

### PR DESCRIPTION
**Pull request by yankinmax**
Backport from: 
 - [ ] https://github.com/OCA/l10n-switzerland/pull/553
 
Depends on
 - [ ] https://github.com/OCA/l10n-switzerland/pull/573

### CODE REVIEW

**Review by i-vyshnevska — COMMENTED** [review]
I would prefer to keep an original history of commits 
to fix build you need to add dependency on edi and your branch
like https://github.com/OCA/account-financial-tools/blob/13.0/oca_dependencies.txt
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#oca_dependencies-txt

**Comment by yankinmax** [comment]
> I would prefer to keep an original history of commits
> to fix build you need to add dependency on edi and your branch
> like https://github.com/OCA/account-financial-tools/blob/13.0/oca_dependencies.txt
> https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#oca_dependencies-txt

Thanks

**Review by grindtildeath — COMMENTED** [review]
LG, just travis setup to fix. Maybe check with @yvaucher

File: l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+    def import_invoice_file(self, file_path, file_name):
+        """Import scanned data from a vendor bill
+
+        And return the created invoice
+        """
+        wiz = self.wiz_import_invoice_file(file_path, file_name)
+        res = wiz.import_invoice()
+        if res.get("res_model") == "account.move":
+            invoice = self.env["account.move"].browse(res["res_id"])
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py]:
```suggestion
        if res.get("res_model") == "account.invoice":
            invoice = self.env["account.invoice"].browse(res["res_id"])
```

File: l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+    def import_invoice_scan(self, invoice_scan):
+        """Import scanned data from a vendor bill
+
+        And return the created invoice
+        """
+        wiz = self.wiz_import_invoice_scan(invoice_scan)
+        res = wiz.import_invoice()
+        if res.get("res_model") == "account.move":
+            invoice = self.env["account.move"].browse(res["res_id"])
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py]: wrong model :fire:

**Review by i-vyshnevska — CHANGES_REQUESTED** [review]

File: l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py (hunk 2 above)
Comment (yankinmax) [reply l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py]: Maybe you know how to solve Travis build issue?
I tried to add python requirements to travis.yml. Didn't help.

**Comment by i-vyshnevska** [comment]
build is done correctly test fails cos there are problems in the code

**Comment by yankinmax** [comment]
> build is done correctly test fails cos there are problems in the code

Will check. Thank you!

File: l10n_ch_qr_bill_scan/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2020 Camptocamp SA
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/__manifest__.py]:
```suggestion
# Copyright 2021 Camptocamp SA
```

File: l10n_ch_qr_bill_scan/models/business_document_import.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+class BusinessDocumentImport(models.AbstractModel):
+    _name = "business.document.import"
+    _description = "Common methods to import business documents"
+
+    @api.model
+    def _hook_match_partner(
+        self, partner_dict, chatter_msg, domain, partner_type_label
+    ):
+        # TODO search by iban
+        return False
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/models/business_document_import.py]:
```suggestion
        return super(BusinessDocumentImport, self)._hook_match_partner(
            partner_dict, chatter_msg, domain, partner_type_label)
```

File: l10n_ch_qr_bill_scan/wizard/account_invoice_import.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        " To enable it, please install pyzbar, pdf2image, cv2 and numpy."
+    )
+    decode = False
+
+
+class AccountInvoiceImport(models.TransientModel):
+    _inherit = "account.invoice.import"
+
+    invoice_scan = fields.Text(string="Scan of the invoice")
+    invoice_file = fields.Binary(string="PDF, PNG or XML Invoice", required=False)
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/wizard/account_invoice_import.py:33]:
```suggestion
    invoice_file = fields.Binary(required=False)
```
if we override args we keeping only those we are overriding

File: l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        cls.env.user.company_id.invoice_import_create_bank_account = True
+        cls.env.ref("base.CHF").active = True
+        cls.supplier = cls.env["res.partner"].create(
+            {
+                "name": "Camptocamp",
+                "street": "EPFL Innovation Park",
+                "street2": "Bldg A",
+                "zip": "1015",
+                "city": "Lausanne",
+                "country_id": cls.env.ref("base.ch").id,
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py:73]:
```suggestion
                "country_id": cls.env.ref("base.ch").id,
                "supplier": True,
```

File: l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            }
+        )
+        cls.supplier.supplier = True
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py]:
```suggestion
```

File: l10n_ch_qr_bill_scan/wizard/account_invoice_import.py (hunk 5 above)
Comment (yankinmax) [reply l10n_ch_qr_bill_scan/wizard/account_invoice_import.py:33]: agree with you! forgot to check original field definition in parent model

Comment (yankinmax) [reply l10n_ch_qr_bill_scan/wizard/account_invoice_import.py:33]: but probably it's also overriding of field's string (we add PNG support)
https://github.com/yankinmax/edi/blob/1ff1ee4c84b3165111de712d97ec85c6a2b142cd/account_invoice_import/wizard/account_invoice_import.py#L23

Comment (i-vyshnevska) [reply l10n_ch_qr_bill_scan/wizard/account_invoice_import.py:33]: ok

**Review by i-vyshnevska — APPROVED** [review]

File: l10n_ch_qr_bill_scan/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+{
+    "name": "Switzerland - QR-bill scan",
+    "summary": "Scan QR-bill to create vendor bills",
+    "version": "12.0.1.0.0",
+    "author": "Camptocamp,Odoo Community Association (OCA)",
+    "category": "Localization",
+    "website": "https://github.com/OCA/l10n-switzerland",
+    "license": "AGPL-3",
+    "depends": ["l10n_ch", "account_invoice_import"],
```
Comment (i-vyshnevska) [inline l10n_ch_qr_bill_scan/__manifest__.py]:
```suggestion
    "depends": ["l10n_ch_qriban", "account_invoice_import"],
```

Comment (yankinmax) [reply l10n_ch_qr_bill_scan/__manifest__.py]: Resolved, thanks

**Comment by yankinmax** [comment]
@yvaucher @i-vyshnevska is it ok, if I cherry-picked here the commits from related PR https://github.com/OCA/l10n-switzerland/pull/573? I tried to add this as oca_dependency, but that way it doesn't work.

**Pull request closed without merging**